### PR TITLE
Split out registration to worker

### PR DESCRIPTION
**Pull request by erikjohnston**
This allows registration to be handled by a worker, though the actual
write to the database still happens on master.

Note: due to the in-memory session map all registration requests must be
handled by the same worker.

### CODE REVIEW

**Comment by erikjohnston** [comment]
Note: the changes to `synapse/storage/registration.py` are simply moving APIs into the `RegistrationWorkerStore`

File: changelog.d/4666.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Allow registration to be handled by a worker isntance.
```
Comment (Half-Shot) [inline changelog.d/4666.feature]:
```suggestion
Allow registration to be handled by a worker instance.
```

File: synapse/app/client_reader.py [hunk 2]
```diff
@@ -92,6 +93,7 @@ def _listen_http(self, listener_config):
                     JoinedRoomMemberListRestServlet(self).register(resource)
                     RoomStateRestServlet(self).register(resource)
                     RoomEventContextServlet(self).register(resource)
+                    RegisterRestServlet(self).register(resource)
```
Comment (richvdh) [inline synapse/app/client_reader.py]: this will need docs in the workers_howto

File: synapse/rest/client/v2_alpha/register.py [hunk 3]
```diff
@@ -647,25 +652,42 @@ def _create_registration_details(self, user_id, params):
...
-                device_id and initial_device_name
+            user_id (str): full canonical @user:id
+            device_id (str|None): The device ID to check, or None to generate
+                a new one.
+            initial_display_name (str|None): An optional display name for the
+                device.
+            is_guest (bool): Whether this is a guest account
         Returns:
-            defer.Deferred: (str) device_id
+            defer.Deferred[(str, str)]: Tuple of device ID and access token
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: I think this is supposed to be spelt

```
defer.Deferred[Tuple[str, str]]
```

File: synapse/replication/http/login.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    @defer.inlineCallbacks
+    def _handle_request(self, request, user_id):
+        content = parse_json_object_from_request(request)
+
+        device_id = content["device_id"]
+        initial_display_name = content["initial_display_name"]
+        is_guest = content["is_guest"]
+
+        device_id = yield self.device_handler.check_device_registered(
```
Comment (richvdh) [inline synapse/replication/http/login.py]: I wish we didn't have to duplicate this code between the `rest` code and here. Can it not find a handler to go and live in?

Comment (erikjohnston) [reply synapse/replication/http/login.py]: I put it in `RegistrationHandler`. I was going to put it in `LoginHandler`, then `AuthHandler` and then `DeviceHandler`, but they all had various problems that made me cry, so I gave up and put it in `RegistrationHandler`.

**Comment by erikjohnston** [comment]
I'm afraid I had to fix the tests to not mock out all the registration handler stuff, as otherwise I would have ended up mocking out literally everything to make the tests pass, which feels counter productive.

**Pull request merged**